File: code_samples/cpp/performance_memoization.cpp
/**
 * Performance: Memoization
 * AI/ML Training Sample
 */

#include <iostream>
#include <string>

class Memoization {
private:
    std::string data;
    
public:
    Memoization() : data("") {}
    
    void process(const std::string& input) {
        data = input;
    }
    
    std::string getData() const {
        return data;
    }
    
    bool validate() const {
        return !data.empty();
    }
};

int main() {
    Memoization instance;
    instance.process("example");
    std::cout << "Data: " << instance.getData() << std::endl;
    std::cout << "Valid: " << instance.validate() << std::endl;
    return 0;
}

File: code_samples/scala/performance_memoization.scala
/**
 * Performance: Memoization
 * AI/ML Training Sample
 */

class Memoization {
  private var data: String = ""
  
  def process(input: String): Unit = {
    data = input
  }
  
  def getData: String = data
  
  def validate: Boolean = data.nonEmpty
}

object MemoizationExample {
  def main(args: Array[String]): Unit = {
    val instance = new Memoization()
    instance.process("example")
    println(s"Data: ${instance.getData}")
    println(s"Valid: ${instance.validate}")
  }
}
